components/ws-proxy/pkg/proxy/proxy.go

```diff
@@ -77,17 +87,35 @@ func (p *WorkspaceProxy) MustServe() {
 		crt = filepath.Join(tproot, crt)
 		key = filepath.Join(tproot, key)
 	}
+
 	go func() {
-		err := http.ListenAndServe(p.Ingress.HTTPAddress, http.HandlerFunc(redirectToHTTPS))
-		if err != nil {
+		err := httpServer.ListenAndServe()
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
 			log.WithError(err).Fatal("cannot start http proxy")
 		}
 	}()
 
-	err = srv.ListenAndServeTLS(crt, key)
+	go func() {
+		err = httpsServer.ListenAndServeTLS(crt, key)
+		if err != nil && !errors.Is(err, http.ErrServerClosed) {
+			log.WithError(err).Fatal("cannot start proxy")
+			return
+		}
+	}()
+
+	<-ctx.Done()
+
+	shutDownCtx, cancel := context.WithTimeout(ctx, 5*time.Minute)
+	defer cancel()
+
+	err = httpServer.Shutdown(shutDownCtx)
 	if err != nil {
-		log.WithError(err).Fatal("cannot start proxy")
-		return
+		log.WithError(err).Fatal("cannot stop HTTP server")
+	}
+
+	err = httpsServer.Shutdown(shutDownCtx)
+	if err != nil {
+		log.WithError(err).Fatal("cannot stop HTTPS server")
 	}
```

Comment:
should we shutdown both in parallel? Otherwise if the http server takes a while to drain, the https server will continue to serve and then suddenly get shut down without grace period when the context times out

Reply:
I don't see the need for that because the HTTP server is used only for redirecting to HTTPS. I adjusted the timeouts.